File: felicia/examples/learn/topic/ros_msg/python/ros_msg_subscribing_node.py
import felicia_py as fel
from felicia.core.protobuf.channel_pb2 import ChannelDef
from felicia.core.protobuf.master_data_pb2 import TopicInfo

from felicia.examples.learn.topic.ros_msg._SimpleMessage import SimpleMessage


class RosMsgSubscribingNode(fel.NodeLifecycle):
    def __init__(self, topic_create_flag):
        super().__init__()
        self.topic = topic_create_flag.topic_flag.value
        self.subscriber = fel.communication.Subscriber(
            SimpleMessage, TopicInfo.ROS)

    def on_did_create(self, node_info):
        self.node_info = node_info
        self.request_subscribe()

    def on_message(self, message):
        print("message : {}".format(message))

    def request_subscribe(self):
        settings = fel.communication.Settings()
        settings.buffer_size = fel.Bytes.from_bytes(512)

        self.subscriber.request_subscribe(self.node_info, self.topic,
                                          ChannelDef.CHANNEL_TYPE_TCP |
                                          ChannelDef.CHANNEL_TYPE_UDP |
                                          ChannelDef.CHANNEL_TYPE_UDS |
                                          ChannelDef.CHANNEL_TYPE_SHM,
                                          settings,
                                          self.on_message)

File: felicia/examples/learn/topic/ros_msg/cc/ros_msg_subscribing_node.cc
#include "felicia/examples/learn/topic/ros_msg/cc/ros_msg_subscribing_node.h"

#include "felicia/core/lib/strings/str_util.h"
#include "felicia/core/master/master_proxy.h"

namespace felicia {

RosMsgSubscribingNode::RosMsgSubscribingNode(
    const TopicCreateFlag& topic_create_flag)
    : topic_create_flag_(topic_create_flag),
      topic_(topic_create_flag_.topic_flag()->value()) {}

void RosMsgSubscribingNode::OnDidCreate(NodeInfo node_info) {
  node_info_ = std::move(node_info);
  RequestSubscribe();
}

void RosMsgSubscribingNode::OnMessage(SimpleMessage&& message) {
  std::cout << "message : " << message << std::endl;
}

void RosMsgSubscribingNode::RequestSubscribe() {
  communication::Settings settings;
  settings.buffer_size = Bytes::FromBytes(512);

  subscriber_.RequestSubscribe(
      node_info_, topic_,
      ChannelDef::CHANNEL_TYPE_TCP | ChannelDef::CHANNEL_TYPE_UDP |
          ChannelDef::CHANNEL_TYPE_UDS | ChannelDef::CHANNEL_TYPE_SHM,
      settings,
      base::BindRepeating(&RosMsgSubscribingNode::OnMessage,
                          base::Unretained(this)));
}

}  // namespace felicia